zaza/utilities/generic.py

```diff
@@ -165,3 +167,219 @@ def get_yaml_config(config_file):
+    # generalized.
+    leader = None
+    non_leaders = []
+    for unit in status["units"]:
+        if status["units"][unit].get("leader"):
+            leader = unit
+        else:
+            non_leaders.append(unit)
+
+    # Pause the non-leaders
+    for unit in non_leaders:
+        if pause_non_leader_subordinate:
+            if status["units"][unit].get("subordinates"):
+                for subordinate in status["units"][unit]["subordinates"]:
+                    logging.info("Pausing {}".format(subordinate))
+                    model.run_action(subordinate, "pause", action_params={})
+        if pause_non_leader_primary:
+            logging.info("Pausing {}".format(unit))
+            model.run_action(unit, "pause", action_params={})
+
+    # Series upgrade the leader
+    logging.info("Series upgrade leader: {}".format(leader))
+    series_upgrade(leader, status["units"][leader]["machine"],
+                   from_series=from_series, to_series=to_series,
+                   origin=origin, workaround_script=workaround_script,
+                   files=files)
+
+    # Series upgrade the non-leaders
+    for unit in non_leaders:
+        logging.info("Series upgrade non-leader unit: {}"
+                     .format(unit))
+        series_upgrade(unit, status["units"][unit]["machine"],
+                       from_series=from_series, to_series=to_series,
+                       origin=origin, workaround_script=workaround_script,
+                       files=files)
+
+
+def series_upgrade(unit_name, machine_num,
+                   from_series="trusty", to_series="xenial",
+                   origin='openstack-origin',
+                   files=None, workaround_script=None):
+    """Perform series upgrade on a unit.
+
+    :param unit_name: Unit Name
+    :type unit_name: str
+    :param machine_num: Machine number
+    :type machine_num: str
+    :param from_series: The series from which to upgrade
+    :type from_series: str
+    :param to_series: The series to which to upgrade
+    :type to_series: str
+    :param origin: The configuration setting variable name for changing origin
+                   source. (openstack-origin or source)
+    :type origin: str
+    :param files: Workaround files to scp to unit under upgrade
+    :type files: list
+    :param workaround_script: Workaround script to run during series upgrade
+    :type workaround_script: str
+    :returns: None
+    :rtype: None
+    """
+    application = unit_name.split('/')[0]
+    juju_utils.prepare_series_upgrade(machine_num, to_series=to_series)
+    logging.info("Watiing for model idleness")
+    model.block_until_all_units_idle()
+    wrap_do_release_upgrade(unit_name, from_series=from_series,
+                            to_series=to_series, files=files,
+                            workaround_script=workaround_script)
+    reboot(unit_name)
+    # Without the sleep model.block_on_all_units_idle returns to early
```

Comment:
I hit the same problem when changing config and added a helper that looks for `executing` status here:
https://github.com/openstack-charmers/zaza/blob/9eda316756f0e4c5f8ddce979f1761f050292cca/zaza/model.py#L653

I guess we could do something similar for this issue at some point

Reply:
Used yours as a model. To update async_block_until_unit_wl_status to actually look at individual unit workload state. Update in the next push.